resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnRMClient.scala

```diff
@@ -93,10 +93,16 @@ private[spark] class YarnRMClient extends Logging {
    * @param status The final status of the AM.
    * @param diagnostics Diagnostics message to include in the final status.
    */
-  def unregister(status: FinalApplicationStatus, diagnostics: String = ""): Unit = synchronized {
-    if (registered) {
-      amClient.unregisterApplicationMaster(status, diagnostics, uiHistoryAddress)
+  def unregister(status: FinalApplicationStatus,
+                 conf: YarnConfiguration,
+                 diagnostics: String = ""): Unit = synchronized {
+    if (!registered) {
+      amClient = AMRMClient.createAMRMClient()
```

Comment:
this is very odd to register during unregister, what if it was the register or start that failed? 
if it timed out and you have a very high timeout I assume you would wait that again.
I'm definitely a bit hesitant about this change.

Reply:
Should a timeout control be added?

Reply:
the question for me is, is this really needed vs possibly introducing other issues.  while it may be convenient, user can just go look in the log to find the error.  You tested a bunch of scenarios which is great, but I'm sure there are others. The main one I mention here is what if registration failed or RM is not responsive?  Can you test out that scenario and see what happens?